File: Word_Break.cpp
/*139. Word Break  QuestionEditorial Solution  My Submissions
Total Accepted: 107015 Total Submissions: 395296 Difficulty: Medium
Given a string s and a dictionary of words dict, determine if s can be segmented into a space-separated sequence of one or more dictionary words.

For example, given
s = "leetcode",
dict = ["leet", "code"].

Return true because "leetcode" can be segmented as "leet code".

Hide Company Tags Google Uber Facebook Amazon Yahoo Bloomberg Pocket Gems
*/



/*DP dynamic programming
use a boolean[] to store if this substring can be find in wordic,
if it is, turn it into true, if not, into false
at the last, check boolean[s.length]
check each substring of s, */

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    bool wordBreak(string s, unordered_set<string>& wordDict) {
        vector<bool> ret(s.size()+1,false);
        ret[0]=true;
        for(int i = 1; i <= s.size(); ++i){//start from 1 to the end of string
            for(int j = 0; j < i; ++j){
                if(ret[j] && wordDict.find(s.substr(j,i-j))!=wordDict.end()){
                    // cout << i << " " << j << endl;
                    ret[i] = true;
                    break;
                }
            }
        }
        return ret[s.size()];
    }
};




/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public boolean wordBreak(String s, Set<String> wordDict) {
        boolean[] ret = new boolean[s.length()+1]; //initialize s.length
        ret[0] = true;
        for(int i = 1; i <= s.length(); ++i){//start from 1 to the end of string
            for(int j = 0; j < i; ++j){
                if(ret[j] && wordDict.contains(s.substring(j,i))){
                    ret[i] = true;
                    break;
                }
            }
        }
        return ret[s.length()];
    }
}

File: Word_Break.swift
/*
Given a string s and a dictionary of strings wordDict, return true if s can be segmented into a space-separated sequence of one or more dictionary words.

Note that the same word in the dictionary may be reused multiple times in the segmentation.

 

Example 1:

Input: s = "leetcode", wordDict = ["leet","code"]
Output: true
Explanation: Return true because "leetcode" can be segmented as "leet code".
Example 2:

Input: s = "applepenapple", wordDict = ["apple","pen"]
Output: true
Explanation: Return true because "applepenapple" can be segmented as "apple pen apple".
Note that you are allowed to reuse a dictionary word.
Example 3:

Input: s = "catsandog", wordDict = ["cats","dog","sand","and","cat"]
Output: false
 

Constraints:

1 <= s.length <= 300
1 <= wordDict.length <= 1000
1 <= wordDict[i].length <= 20
s and wordDict[i] consist of only lowercase English letters.
All the strings of wordDict are unique.
*/

/*
Solution 2:
DP

Time Complexity: O(n^3)
Space Complexity: O(n)
*/
class Solution {
    func wordBreak(_ s: String, _ wordDict: [String]) -> Bool {
        let n = s.count
        var s = Array(s)
        var wordDict = Set(wordDict)
        
        var dp = Array(repeating: false, count: n+1)
        dp[0] = true
        
        for i in 1...n {
            for j in 0..<i {
                if dp[j] && wordDict.contains(String(s[j..<i])) {
                    dp[i] = true
                    break
                }
            }
        }
        return dp[n]
    }
}


/*
Solution 1:
BFS with memoization

Time Complexity: O(n^3)
Space Complexity: O(n)
*/
class Solution {
    func wordBreak(_ s: String, _ wordDict: [String]) -> Bool {
        let n = s.count
        var s = Array(s)
        
        var wordDict = Set(wordDict)
        var queue: [Int] = [0]
        var visited = Array(repeating: false, count: n)
        while !queue.isEmpty {
            let start = queue.removeFirst()
            if visited[start] { continue }
            
            visited[start] = true
            for end in (start+1)...n {
                if wordDict.contains(String(s[start..<end])) {
                    queue.append(end)
                    if end == n { 
                        return true
                    }
                }
            }
        }
        return false
    }
}
